cpp/src/arrow/filesystem/azurefs_test.cc

```diff
@@ -451,10 +470,14 @@ TEST_F(AzureHierarchicalNamespaceFileSystemTest, CreateDirRecursiveSuccessDirect
   arrow::fs::AssertFileInfo(fs_.get(), parent, FileType::Directory);
 }
 
-TEST_F(AzuriteFileSystemTest, CreateDirRecursiveFailureDirectoryOnly) {
+TEST_F(AzuriteFileSystemTest, CreateDirRecursiveSuccessDirectoryOnly) {
   const auto parent = PreexistingContainerPath() + RandomDirectoryName();
   const auto path = internal::ConcatAbstractPath(parent, "new-sub");
-  ASSERT_RAISES(NotImplemented, fs_->CreateDir(path, true));
+  ASSERT_OK(fs_->CreateDir(path, true));
+  // There is only virtual directory without hierarchical namespace
+  // support. So the CreateDir() does nothing.
+  arrow::fs::AssertFileInfo(fs_.get(), path, FileType::NotFound);
+  arrow::fs::AssertFileInfo(fs_.get(), parent, FileType::NotFound);
```

Comment:
If we can disambiguate "file exists?" queries in the filesystem API, we should probably always reply true when the caller is asking if a directory exists. If creating a directory is a no-op that succeeds, the post-condition of CreateDir -- the directory now exists -- should be true.

There are might be bad consequences of this, so this is more of an idea than a suggestion.

Reply:
I don't feel strongly but personally I think the current behaviour is the best option. 

`GetFileInfo` will return that a directory is present if at least one "file" has been created in that "directory". I think this behaviour is consistent with the GCS filesystem.

Reply:
I agree with the opinion of Tom-Newton but let's discuss this further on #38772.